_drafts/unit-testing-with-mockito.md

```diff
@@ -0,0 +1,129 @@
+---
+layout: post
+title:  "Unit Testing with Mockito for Android"
+author: johnleeroy
+description: Learn how to unit test your Android code with JUnit4 and Mockito.
+categories: 
+  - android
+  - java 
+  - unit testing
+  - tdd
+  - code coverage
+  - tech
+---
+Every developer knows that testing software can take a lot of time.  Being the brilliant industrious individuals that we are, engineers invented
+multiple ways to test software.  Unit testing is just one part among others such as functional testing, integration testing, and acceptance testing.
+In this blog, we will go over how to write unit tests for Android using the latest tools and libraries.  In this example, I am leverage JUnit4 and Mockito.  
+
+# What is Unit Testing?
+Tests the smallest functionality of a feature.  This is generally a function with a single responsibility.  Unit tests shouldn't be concerned with outside dependancies and if they exist in the they, code should be mocked or stubbed.  You generally test the public interface of your code which should call on any and all unaccessible implementation hidden within the class.
+
+The goal is to create a test that is relatively simple, easy to debug, fast to execute, and validates that your unit of functionality works within it's own encapsulation before it is used by another object.
+
+# Example : Buying a Drink
+### Scenario
+We have a Patron, Matt, who enters a cafe and buys a cold, refreshing Nuka Cola.  To unit test the functionality of buying a drink, we will mock both the Cafe and DrinkInterface because the implementation of these objects are largely irrelevant to testing the Patron class.  
+
+##### Patron (abridged)
+```java
+public class Patron {
+
+    private String mName;
+    protected float mCash;
+    protected Cafe mCurrentCafe;
+    protected ArrayList<DrinkInterface> mDrinksInPossession;
+
+    public Patron(String name, float cashAmount) {
+        mName = name;
+        setCash(cashAmount);
+        mDrinksInPossession = new ArrayList<>();
+    }
+
+    public void enterCafe(Cafe cafe) {
+        cafe.addNewCustomer(this);
+        mCurrentCafe = cafe;
+    }
+
+    public void purchaseDrinkFromCafe(String drinkName) throws NotEnoughMoneyException {
+        DrinkInterface drink = mCurrentCafe.getDrink(drinkName);
+        float cost = drink.getCost();
+        if (mCash >= cost) {
+            mCash -= cost;
+        } else {
+            return null;
+        }
+        DrinkInterface newDrink = mCurrentCafe.sellDrink(drink.getName());
+        mDrinksInPossession.add(newDrink);
+    }
+}
+```
+##### DrinkInterface
+```java
+public interface DrinkInterface {
+    String getName();
+    float getCost();
+    float getSizeInOz();
+
+    abstract DrinkInterface getCopy();
+}
+```
+
+### Setup
+We create Matt to test the Patron class.  
+PatronHelper is a child of the Patron class which helps reveal unaccesable properties for our tests.  
```

Comment:
Let's wrap things that refer to code with quotes.. so here we could use something like `PatronHelper` instead.